falcon/middlewares.py

```diff
@@ -1,16 +1,86 @@
+from typing import Iterable, Optional, Union
+
+from .request import Request
+from .response import Response
+
 
 class CORSMiddleware(object):
-    def process_response(self, req, resp, resource, req_succeeded):
-        """Implement a simple blanket CORS policy for all routes.
+    """CORS Middleware
+
+    This middleware provides a simple out-of-the box CORS policy, including handling
+    of preflighted requests from the browser.
+
+    See also:
+
+    * https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+    * https://www.w3.org/TR/cors/#resource-processing-model
+
+    Keyword Arguments:
+        allow_origin (Union[str, Iterable[str]]): List of origins to allow (case
+            sensitive). The string ``'*'`` acts as a wildcard, matching every origin.
+            (default ``'*'``).
+        expose_headers (Optional[Union[str, Iterable[str]]]): List of additional headers to
+            expose via the ``Access-Control-Expose-Headers`` header. These headers are in addition
+            to the CORS-safelisted ones: ``Cache-Control``, ``Content-Language``,
+            ``Content-Length``, ``Content-Type``, ``Expires``, ``Last-Modified``, ``Pragma``.
+            (default ``None``).
+
+            See also:
+            https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers
+        allow_credentials (Union[bool, Iterable[str]]): List of origins to allow credentials (case
```

Comment:
It's totally a matter of taste, but I don't think we use this pattern much elsewhere in Falcon... ("Polymorphic" union parameters which can be ``False``, ``True`` and something else.)

Have you considered splitting into explicit parameters, probably something along the lines of how [`falcon-cors` does it](https://falcon-cors.readthedocs.io/en/master/falcon_cors.html)?

Reply:
I did think about splitting it, but I think it may be more convenient this way. After parsing it has only two states: True or a set of string (empty when this argument is false), so I don't think it overly complicates matters.

Also with two parameters we would need to decide how they interact, who takes precedence and how the bool parameter is used. Does True enable all origins or when used with a list of string the list takes precedence? Does False disable all origins or the list takes precedence? A single parameter provides a cleaner interface in my opinion.

Reply:
I wanted to update this. I still prefer the single argument, but I can change this to enable two arguments.
@vytas7 do you have suggestions on what interaction should the two arguments have? In falcon-cors the `allow_credentials_all_origins` takes precedence if `True`, else `allow_credentials_origins_list` is used if the `allow_credentials_all_origins` was `False`.

@kgriffs do you have preferences?

Reply:
The single-argument approach is reminiscent of how things often work in the Node.js community; it can be a bit strange until you get used to it. But I can't really think of an alternative which would be better since it may actually be harder to reason about two params, and it would be more of a burden to pass a dict, or set of tuples, to `allow_origin` that explicitly provides `True`/`False` for each origin. 

So I am leaning toward just keeping it as-is, but perhaps making the docstring a little more clear, for example:

```rst
List of origins for which to allow credentials (case sensitive). Alternatively, a boolean
value may be passed to specify that credentials should be allowed for all allowed 
origins (``True``) or disallowed for all (``False``). This parameter takes effect only 
if the origin is allowed by the ``allow_origin`` argument. (Default ``False``).
```

Reply:
Since this is just some batteries included, and not a part of the core API, I'm fine with it either way.

I'm still a bit uneasy with the values this union parameter accepts though, particularly in the light of `expose_headers` accepting `None` (and it is even the default value).
Regardless of how we document it, this may be confusing: `expose_headers=None` -- none -- no headers to expose -- sounds logical, but passing `None` to `allow_credentials` (allone none -- allow no credentials one may think) would explode while trying to make a `frozenset` out of `None`?

With that in mind, would it be worth making the first check just `not allow_credentials` or more explicitly allow `is None`?

Reply:
Sure we can make that change. I'll have another look and try to normalize the parameters a bit incorporating the suggestions.